### PR TITLE
Script and document Windows support (rebased on v2.0)

### PR DESCRIPTION
**Pull request by sneurlax**
This PR adds basic Windows support.

See the `README.md` changes and the commit history for a summary of the changes.  Basically, I used `flutter create . --platforms=windows`, added a `build_windows.dart` build script, a `crosscompile_windows.dart` Dockerfile-based script, and a `build_wsl.dart` to be ran on a Windows host in a WSL2 instance, and I documented their use.

 - [ ] Implement [requested changes](https://github.com/peercoin/coinlib/pull/23/files/4ee47ce066ebe6e054cea81e39e7d5a108ae7f73).
   - [x] Move dependencies from script to docs.
   - [x] Move DLL to build/secp256k1.dll IAW lib/src/secp256k1/secp256k1_io.dart.
   - [x] `processSync`->`execWithStdio`.
   - [x] Remove unnecessary autogenerated files.
   - [x] Use temporary directory in order to avoid polluting project directory.
   - [x] Put DLL in `/build` (vs. `/build/windows` or `/build/windows/x64`)
   - [ ] Integrate build process into the flutter build process (see [discussion](https://github.com/peercoin/coinlib/pull/23#discussion_r1431708795)).
 - [x] Cross-compile for Windows using dockerfile
 - [x] Cross-compile for Windows using WSL
 - [x] Native Windows build
 - [x] Update PR text, [referring to previous PR](https://github.com/peercoin/coinlib/pull/23/files/4ee47ce066ebe6e054cea81e39e7d5a108ae7f73).
 - [x] Mark PR ready for review.

### CODE REVIEW

**Comment by sneurlax** [comment]
I marked the PR ready for review while working on the CMake integration on Windows.  It's ready on a basic level as-is but [could use improvements as discussed](https://github.com/peercoin/coinlib/pull/23#discussion_r1431708795) on the [previous PR](https://github.com/peercoin/coinlib/pull/23#discussion_r1431678890).  I'm not sure that full Flutter CMake integration will be feasible for me on Windows, but will come back with an update on this point in awhile.

File: coinlib/bin/build_wsl.dart
Comment (MatthewLM) [inline coinlib/bin/build_wsl.dart:42]: I suppose this was for debugging purposes and no longer needed?

File: .gitignore
Comment (MatthewLM) [inline .gitignore:3]: This shouldn't be necessary any more.

File: coinlib_flutter/windows/CMakeLists.txt
Comment (MatthewLM) [inline coinlib_flutter/windows/CMakeLists.txt:13]: This is where the CMakeLists.txt for Linux/Android is included. I'm not sure if this can be used for Windows too or if this should be removed in favour of Windows specific configuration placed into this file.

File: coinlib/bin/build_windows.dart
Comment (MatthewLM) [inline coinlib/bin/build_windows.dart:45]: Double newline here.

**Review by MatthewLM — CHANGES_REQUESTED** [review]
Thank you for these changes. There's only a few minor comments.

Hopefully it is possible to integrate the windows build into flutter with CMake. If not, then the `coinlib_flutter` README can notify consumers of the plugin that the DLL has to be built using the script before a successful flutter build is possible.

File: coinlib/bin/build_wsl.dart
Comment (sneurlax) [reply coinlib/bin/build_wsl.dart:42]: Oh, yes, absolutely.

File: coinlib/bin/build_windows.dart
Comment (sneurlax) [reply coinlib/bin/build_windows.dart]: Resolved by 6f620a476fcce8afb87fa35c90f81a341d886abb. `build_linux.dart` affected, as well

File: coinlib/bin/build_wsl.dart
Comment (sneurlax) [reply coinlib/bin/build_wsl.dart]: Resolved by 6f620a476fcce8afb87fa35c90f81a341d886abb

File: .gitignore
Comment (sneurlax) [reply .gitignore]: Resolved by c416cd3efa87a3f0d396ef3ae5df46559088e385

File: coinlib_flutter/windows/CMakeLists.txt
Comment (sneurlax) [reply coinlib_flutter/windows/CMakeLists.txt:13]: My issues with implementing this today have mostly revolved around the disconnect between needing to run Flutter on the Windows host, but only being able to build in a WSL instance on Windows--what builds it can't run it, what runs it can't build it.

I may be able to build with native Windows tools.  Maybe there's a workaround or other approach I could take to accomplish the task.  But as for today, I wasn't able to complete the work as requested, sorry.

Comment (MatthewLM) [reply coinlib_flutter/windows/CMakeLists.txt:13]: No problem! I can leave it to you if this is something you'd like to continue with. If not, then that's fine, we can instead require users build the library separately. Either way, I appreciate the contribution. For Flutter users it is good to have some way to build for Windows, even if it's not as seamless as the other platforms.

With little time before Christmas, I can hold off merging. I'm happy to come back to this in the new year.

Comment (sneurlax) [reply coinlib_flutter/windows/CMakeLists.txt:13]: I'll try again before the new year, but for now I have to work on integrating this currently-working (but manually-built) solution into other projects ([fusiondart](https://github.com/cypherstack/fusiondart) and [flutter_libsparkmobile](https://github.com/cypherstack/flutter_libsparkmobile/tree/testing)).

Late last night I thought that maybe I could use the Ubuntu WSL2 instance to navigate to the flutter project to be built/ran on Windows and run a `flutter run`/`flutter build`, which would cause CMake to cross-compile the lib before it will probably fail to complete the build or at least run it.  That might be a working approach.  But it's still a somewhat manual process...  Or I can try to get it building on Windows natively, though that comes with additional Visual Studio requirements, most likely.  That might be mostly-automatic except for requiring the right dependencies.  Would either of those approaches work, or would you prefer the latter?

Comment (MatthewLM) [reply coinlib_flutter/windows/CMakeLists.txt:13]: I imagine if users require the WSL2, it's almost as easy to run the build script than use an additional `flutter build` command? I think it would be reasonable to just keep the build script in that case. If it can instead be done natively through a single invocation of the flutter build process whilst requiring some dependencies I think that would be a good solution.

I'll be around for a while tomorrow and I might be around next Wednesday and/or Saturday. Though it might be that I'll have to take a look at any further changes in January.

**Comment by sneurlax** [comment]
The coinlib example app was running yesterday and the day before, but it isn't for me currently, so unfortunately I must retract this PR :( I may need to enlist another Windows user for independent testing

**Comment by sneurlax** [comment]
I resolved the issue with the coinlib example on Windows, so I'm re-marking this PR as ready to review, sorry for the switcheroo.

Note that I had to alter `coinlib_flutter/example/pubspec.yaml` like
```
diff --git a/coinlib_flutter/example/pubspec.yaml b/coinlib_flutter/example/pubspec.yaml
index c06782d..2a7fccc 100644
--- a/coinlib_flutter/example/pubspec.yaml
+++ b/coinlib_flutter/example/pubspec.yaml
@@ -10,7 +10,12 @@ environment:
 dependencies:
   flutter:
     sdk: flutter
-  coinlib_flutter: ^1.0.0
+  coinlib_flutter:
+    git:
+      url: https://github.com/cypherstack/coinlib.git
+      path: coinlib_flutter
+      ref: windows # TODO change to main.  TODO update to windows-2.0 branch.
+      # TODO pin to specific commit to reduce supply chain attack surface. 
   cupertino_icons: ^1.0.2
 
 dev_dependencies:
```
however this shouldn't be an issue if/after the changes are published to pub.dev

File: coinlib/lib/src/secp256k1/secp256k1_io.dart
Comment (MatthewLM) [inline coinlib/lib/src/secp256k1/secp256k1_io.dart:23]: Is this `windows/` folder/prefix required to load on flutter? If so, why not include it for the `flutterLib` but keep the pure-dart library location in `build/` only to keep consistency with the other builds?

File: coinlib_flutter/windows/CMakeLists.txt
Comment (MatthewLM) [inline coinlib_flutter/windows/CMakeLists.txt:24]: If it's not possible to build the library using CMake as part of the flutter process, then it needs to pick up the library from the PWD where the build script was run. I'm not sure how to get that in the CMake flutter process.

Alternatively the binary could be hosted and downloaded.

**Review by MatthewLM — CHANGES_REQUESTED** [review]
Thank you for these continued changes. Is there any hope to get the Windows CMake build working? Otherwise the library as built by the user has to be picked up by CMake.

At some point I can make an experimental windows release to check that it works as a package such as `2.0.0-win-beta.1`.

File: coinlib/lib/src/secp256k1/secp256k1_io.dart
Comment (sneurlax) [reply coinlib/lib/src/secp256k1/secp256k1_io.dart:23]: Yes, the addition of the `windows` subfolder was required in order to use `coinlib_flutter` in another flutter project.  coinlib_flutter's own example runs fine in either location, but my consumer needed it in `build/windows/x64`

File: coinlib_flutter/windows/CMakeLists.txt
Comment (sneurlax) [reply coinlib_flutter/windows/CMakeLists.txt:24]: This is testing working as part of the coinlib_flutter example app flutter build process.

**Comment by sneurlax** [comment]
> Is there any hope to get the Windows CMake build working?

Yes, I don't feel that I exhausted every approach towards that end.  It could very well be viable.  I will try to try again before the New Year, but have been reassigned elsewhere for the time being

File: coinlib_flutter/windows/CMakeLists.txt
Comment (MatthewLM) [reply coinlib_flutter/windows/CMakeLists.txt:24]: If someone is using the package from pub.dev, the library wont be in the `CMAKE_BINARY_DIR` though right? The library needs to be picked up from the consumer's local build.

File: coinlib/lib/src/secp256k1/secp256k1_io.dart
Comment (MatthewLM) [reply coinlib/lib/src/secp256k1/secp256k1_io.dart:23]: OK, can't the `localLib` be without the `windows` folder though?

**Comment by MatthewLM** [comment]
> Yes, I don't feel that I exhausted every approach towards that end. It could very well be viable. I will try to try again before the New Year, but have been reassigned elsewhere for the time being

OK, no worries. I might be unavailable until the latter part of next week, but I will take another look at anything when I can.

**Comment by sneurlax** [comment]
>  If someone is using the package from pub.dev, the library wont be in the
CMAKE_BINARY_DIR though right?

Hmm, I'm not sure about that, actually!  I tested it using both path and
git refs locally (uncommitted).  First I made sure the coinlib_flutter
example works, then I made sure the whole setup works when included in
another flutter project.  When I `dart run coinlib:build_wsl`, it spits the
secp256k1.dll into <flutter_project>/build/windows and coinlib finds it
there as well, so I think `CMAKE_BINARY_DIR` is valid here.

But it would probably be a good idea to put out a windows testing version
as you mentioned before before adding it to a "mainstream" release.

**Comment by sneurlax** [comment]
I'll have to test this again, I thought this change was consistent with the
.dll placement in build/windows.

File: coinlib/lib/src/secp256k1/secp256k1_io.dart
Comment (sneurlax) [reply coinlib/lib/src/secp256k1/secp256k1_io.dart:23]: I'll have to test this again, I thought this change was consistent with the .dll placement in build/windows.

File: coinlib_flutter/windows/CMakeLists.txt
Comment (sneurlax) [reply coinlib_flutter/windows/CMakeLists.txt:24]: I think CMAKE_BINARY_DIR  would be `<flutter_project>/build/windows/x64 (hence the `/../`).  I tested it using both path and git refs locally (uncommitted).  First I made sure the coinlib_flutter example works, then I made sure the whole setup works when included in another flutter project.  When I `dart run coinlib:build_wsl`, it spits the `secp256k1.dll` into `<flutter_project>/build/windows` and coinlib finds it there as well, so I think `CMAKE_BINARY_DIR` is valid here.

But it would probably be a good idea to put out a windows testing version as you mentioned before before adding it to a "mainstream" release to doublecheck that this is the case when using it from pub.dev.

**Comment by sneurlax** [comment]
I'm working on a native Windows build now.  This might be counting my chickens before they hatch, but I wonder about the build scripts names...  Right now:

 - `build_windows.dart` cross-compiles for Windows from a Linux host using docker/podman
 - `build_wsl.dart` cross-compiles for Windows from a WSL instance on a Windows host
 
Do you have any preference how these should be named it I can accomplish a native Windows build?  I'd assume `build_windows.dart` would be for running *on* Windows, personally.  I don't know how I'd rename the cross compilation script(s), do you have any preferences there, assuming I do manage to achieve a native Windows build?

**Comment by sneurlax** [comment]
I have native Windows builds working, but not integrated into coinlib_flutter/src/CMakeLists.txt yet.

File: coinlib/lib/src/secp256k1/secp256k1_io.dart
Comment (sneurlax) [reply coinlib/lib/src/secp256k1/secp256k1_io.dart:23]: Resolved by 91f0f7f0e66e6ded1f58715ba5d5653d2fbb9402

Puts secp256k1.dll in /build instead of /build/windows or /build/windows/x64

File: coinlib_flutter/windows/CMakeLists.txt
Comment (sneurlax) [reply coinlib_flutter/windows/CMakeLists.txt]: I was not able to get this completely working using just the CMakeLists.  `build_windows.dart` does work natively on Windows, however.

File: coinlib_flutter/windows/CMakeLists.txt
Comment (sneurlax) [reply coinlib_flutter/windows/CMakeLists.txt]: This conversation has been outdated by 91f0f7f0e66e6ded1f58715ba5d5653d2fbb9402 but still applies broadly.  I still think it's correct, do you have a suggested change?

**Comment by sneurlax** [comment]
OK, I have not been able to get the entire build working automagically on Windows by altering `coinlib_flutter/src/CMakeLists.txt`, but I have native Windows builds working and documented.  This is in a good place!  It could be better by just using CMake and not `coinlib/coinlib/bin/build_windows.dart` *et al.*, but I wasn't able to accomplish that yet.  Perhaps that's a good topic for some future issues?  I have a feeling that Linux could be made automagic...

File: coinlib_flutter/README.md
Comment (MatthewLM) [inline coinlib_flutter/README.md:29]: "For desktop platforms" ought to be "For Windows". Perhaps here direct instruction on building the library can be included. I guess `dart run coinlib:build_windows` (or wsl) should be run and then the flutter CMake build will pick up the library from the local build folder?

File: coinlib_flutter/example/pubspec.yaml
Comment (MatthewLM) [inline coinlib_flutter/example/pubspec.yaml:5]: How come this is changed to 2.1.0?

File: coinlib_flutter/windows/CMakeLists.txt
Comment (MatthewLM) [inline coinlib_flutter/windows/CMakeLists.txt:24]: How is the library being picked up with this? Surely once it is built, the library would be in `${CMAKE_BINARY_DIR}/build/secp256k1.dll`?

**Review by MatthewLM — COMMENTED** [review]
Many thanks for these changes. I've got some questions in this review before I go ahead and make a test pre-release.

File: coinlib/README.md
Comment (sneurlax) [reply coinlib/README.md]: Resolved by 83a734985caca8b0534fb72891924540d5508791

File: coinlib_flutter/README.md
Comment (sneurlax) [reply coinlib_flutter/README.md]: Resolved by 8d350c26ae91a2a355e5d1e7e2e189582428c273

Is 2.0 automatically built for linux et al?  We're using 1.0 right now and I thought we had to construct coinlib for the other desktop platforms.

File: coinlib_flutter/windows/CMakeLists.txt
Comment (sneurlax) [reply coinlib_flutter/windows/CMakeLists.txt:24]: `CMAKE_BINARY_DIR` is `app/build/windows/x64`.  The DLL is copied into `app/build`.  There's probably another CMAKE variable available that refers to just `app/build`, but this works.

File: coinlib_flutter/example/pubspec.yaml
Comment (sneurlax) [reply coinlib_flutter/example/pubspec.yaml:5]: I chose 2.1.0 because I assumed the example app would be bumped to 2.0 for v2.0, a 2.1 because this is 2.0 + windows (the example windows CMakeLists are altered).  It can be changed to anything, though, what would you prefer?  No change?

Comment (MatthewLM) [reply coinlib_flutter/example/pubspec.yaml:5]: I've left it as it is as it's only an example app. It doesn't matter what the version is really.

Comment (MatthewLM) [reply coinlib_flutter/example/pubspec.yaml:5]: And, there is possibility of the Windows support being added to v2.0. It depends on when a final release is made. I'm still waiting on feedback for the Taproot changes.

File: coinlib_flutter/windows/CMakeLists.txt
Comment (MatthewLM) [reply coinlib_flutter/windows/CMakeLists.txt:24]: OK thank. It will be good to verify this works with the test package. I assume everything is ready to test that? I might be able to publish a test release this weekend and I can see if there are a few others who can test.

File: coinlib_flutter/example/pubspec.yaml
Comment (sneurlax) [reply coinlib_flutter/example/pubspec.yaml:5]: Well please feel free to change it to whatever needed.  I'm more than happy to change it and push before merge, just let me know!

File: coinlib_flutter/windows/CMakeLists.txt
Comment (sneurlax) [reply coinlib_flutter/windows/CMakeLists.txt:24]: I considered moving the lib to CMAKE_BINARY_DIR but thought it would be better to remain consistent with coinlib conventions across platforms

File: coinlib_flutter/example/pubspec.yaml
Comment (MatthewLM) [reply coinlib_flutter/example/pubspec.yaml:5]: I'd say keep it as 1.0.0 though it's not very important.

File: coinlib_flutter/windows/CMakeLists.txt
Comment (sneurlax) [reply coinlib_flutter/windows/CMakeLists.txt:24]: > OK thank. It will be good to verify this works with the test package. I assume everything is ready to test that? I might be able to publish a test release this weekend and I can see if there are a few others who can test.

Yes, everything is ready to test! 🥳

**Comment by sneurlax** [comment]
Reverted/resolved by 50c49e173feb12068a3cb8007e1c87dd522a16ba, example app
version is 1.0.0

**Comment by MatthewLM** [comment]
@sneurlax I merged a squashed commit to a new windows branch in preparation for a test pre-release. I'll likely get to that on Sunday ready for testing next week. Many thanks.

**Pull request merged**
